File: provider/src/protocol.rs
//! Wire types for advisor ↔ provider WebSocket messages.
//!
//! The advisor is stateless w.r.t. receipts (anything that matters for
//! authority lives on a PDS), but we still need a transport for: provider
//! registration, heartbeats, encrypted job dispatch, and inference
//! progress. Frames are tagged JSON (`#[serde(tag = "type", rename_all =
//! "snake_case")]`) so the wire is human-debuggable and version-tolerant —
//! receivers can ignore unknown variants without dropping the connection.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AdvisorMessage {
    /// Provider → advisor: announce capabilities.
    Register(Register),
    /// Provider → advisor: liveness and load.
    Heartbeat(Heartbeat),
    /// Advisor → provider: encrypted job to serve.
    InferenceRequest(InferenceRequest),
    /// Provider → advisor: streaming response chunk.
    InferenceChunk(InferenceChunk),
    /// Provider → advisor: "still generating" signal sent during a long job
    /// when no user-visible token has gone out for a while (slow prefill or
    /// a slow decode patch). Resets the advisor's session idle timer without
    /// being a token, so a slow-but-alive job isn't killed as silent.
    /// Additive — an old advisor ignores the unknown frame.
    InferenceKeepalive(InferenceKeepalive),
    /// Provider → advisor: terminal completion notice.
    InferenceComplete(InferenceComplete),
    /// Advisor → provider: nonce-based attestation challenge.
    AttestationChallenge(AttestationChallenge),
    /// Provider → advisor: signed challenge response.
    AttestationResponse(AttestationResponse),
    /// Provider → advisor (→ requester): per-request ephemeral session key for
    /// the confidential tier. Minted fresh inside the measured engine and
    /// SE-signed over the requester's nonce + the active attestation CID, so a
    /// confidential requester can prove the key it seals to is controlled by
    /// the attested enclave and was produced for THIS request (not replayed).
    SessionKey(SessionKey),
    /// Advisor → provider: liveness probe. The provider answers `Pong`
    /// with the same nonce immediately from its serve loop, so a pong
    /// proves the request loop is pumping (stronger than a WS-level pong,
    /// which the read half auto-sends even when the loop is wedged). The
    /// advisor uses this to preflight a provider before routing a job.
    Ping(Ping),
    /// Provider → advisor: reply to `Ping`, echoing the nonce.
    Pong(Pong),
    /// Provider → advisor: response to an APNs code-identity challenge.
    /// The challenge is delivered out-of-band over APNs (not on this socket):
    /// the advisor pushes a nonce sealed to the provider's X25519 key `K`, and
    /// only the genuine, Apple-provisioned binary can (a) receive that push at
    /// all (AMFI gates the topic to our code signature) and (b) decrypt it with
    /// `K`. The provider recovers the nonce and returns it here alongside a
    /// Secure-Enclave signature over it, proving the same binary that received
    /// the push also controls the attested signing key. Additive — pre-APNs
    /// advisors never send a challenge and so never receive this.
    CodeAttestationResponse(CodeAttestationResponse),
    /// Advisor → provider: standing change. `Bad` means the advisor
    /// stopped routing jobs here (failed preflight / went silent mid-job);
    /// the agent surfaces it to the operator. `Ok` clears it. Additive.
    HealthNotice(HealthNotice),
    /// Advisor → provider: "your owner-set control state changed — re-read
    /// it." A pure nudge carrying no authority; the agent re-reads the
    /// authoritative `active` from its own PDS record and reports it back
    /// in a Heartbeat. Lets a console start/stop take effect in ~a second
    /// instead of waiting for the next poll. Additive.
    ControlChanged(ControlChanged),
    /// Advisor → provider: "you've been flagged unhealthy — try to
    /// self-right now." The agent runs its engine health-check + bounded
    /// restart immediately (rather than waiting for the next scheduled
    /// health tick) and reports the outcome in `RecoverResult`. Carries no
    /// authority. Additive — old agents ignore it (they still get the
    /// `HealthNotice` that accompanies it).
    RecoverRequest(RecoverRequest),
    /// Provider → advisor: outcome of a self-right attempt. `recovered`
    /// true → engines are back and the machine is ready for jobs again
    /// (advisor clears its bad standing); false → it couldn't, and `detail`
    /// is the remediation text the console + tray surface. Additive.
    RecoverResult(RecoverResult),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecoverRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RecoverResult {
    pub recovered: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ControlChanged {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Ping {
    pub nonce: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Pong {
    pub nonce: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HealthStanding {
    Ok,
    Bad,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthNotice {
    pub standing: HealthStanding,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Register {
    pub provider_did: String,
    /// Stable per-machine identifier — this machine's
    /// `dev.cocore.compute.provider` record rkey. Lets the advisor hold
    /// several machines under one DID (an owner serving on a laptop AND a
    /// desktop) instead of one evicting the other, and is the join key the
    /// console uses to map advisor live-standing onto a machine. Optional /
    /// additive: pre-`machine_id` advisors ignore it, and an advisor that
    /// wants it but doesn't get it falls back to `attestation_pub_key`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub machine_id: Option<String>,
    pub machine_label: String,
    pub chip: String,
    pub ram_gb: u32,
    pub supported_models: Vec<String>,
    pub encryption_pub_key: String,  // base64 X25519
    pub attestation_pub_key: String, // base64 P-256
    pub attestation_uri: String,     // at://… of the active attestation record
    /// Present when the agent could not bring one or more configured
    /// inference engines online after its startup recovery loop gave up
    /// (see `build_engines`). Purely diagnostic: the failed models are
    /// already absent from `supported_models`, so matchmaking is correct
    /// without this — it exists so the advisor can *note* that a machine
    /// is up but degraded (only serving `stub`) and surface that to
    /// operators / dashboards. Omitted on a healthy serve. Additive: old
    /// advisors ignore it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub engine_fault: Option<crate::pds::EngineFault>,
    /// Measured cdHash of the running binary, echoed from the signed
    /// attestation so the advisor can check it against its known-good set when
    /// computing confidential eligibility (an accelerator only — a confidential
    /// requester re-verifies the PDS attestation at seal time). Additive.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cd_hash: Option<String>,
    /// Provider's self-asserted tier (`attested-confidential` | `best-effort`),
    /// echoed from the signed attestation. Advisory — the advisor recomputes
    /// eligibility from cdHash ∈ known-good + challenge-verified SIP. Additive.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tier: Option<String>,
    /// Coarse, opt-in ISO 3166-1 alpha-2 country echoed from the provider
    /// record's `region`, so the advisor's `/providers` list can route by
    /// country without reading the PDS. Advisory self-claim; `None` when the
    /// owner hasn't opted into location sharing. Additive.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub region: Option<String>,
    /// APNs device token (hex) for this machine's measured agent process, when
    /// it could register for remote notifications (a logged-in GUI session with
    /// the push entitlement). Lets the advisor send the APNs code-identity
    /// challenge that proves this exact binary is the genuine, team-signed
    /// agent — the un-self-reportable complement to `cd_hash`. Omitted on
    /// headless/launchd installs (no GUI session → no APNs), which therefore
    /// stay best-effort. Additive — pre-APNs advisors ignore it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub apns_device_token: Option<String>,
    /// Version string of this agent binary (`env!("CARGO_PKG_VERSION")`,
    /// e.g. `0.9.32`), echoed live so the advisor can route version-gated
    /// jobs — a request that needs a feature only present from some release
    /// (e.g. `messages-v1` image input) is steered to machines at/above a
    /// minimum version. Mirrors the `binaryVersion` already stamped on the
    /// PDS provider record. Additive: pre-version advisors ignore it, and an
    /// advisor enforcing a floor treats a machine that omits it as below the
    /// floor (fail-closed).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub binary_version: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Heartbeat {
    pub load: f32,
    pub queue_depth: u32,
    pub at: chrono::DateTime<chrono::Utc>,
    /// The owner's start/stop switch as the agent currently sees it on its
    /// PDS record. `Some(false)` → the owner stopped this machine; the
    /// advisor routes it no jobs. `Some(true)` / absent → serving. Additive
    /// — old advisors ignore it (and the machine keeps serving).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub active: Option<bool>,
    /// Content-free crash telemetry: how many times this machine's agent
    /// has panicked since its last sustained clean uptime, plus the most
    /// recent panic's location/signature/build. Lets the advisor (and an
    /// operator dashboard) spot a flapping machine fleet-wide without
    /// anyone tailing a log — the symptom that previously only showed up
    /// as a quietly-flat ledger. Carries NO prompt/token content and NO
    /// panic message (only a `file:line` and a hash). Omitted on a machine
    /// that has never crashed. Additive — old advisors ignore it.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub crash: Option<crate::diagnostics::CrashSignature>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InferenceRequest {
    pub job_uri: String, // at:// strong-ref to the job record
    /// CID half of the job strong-ref. Optional on the wire because
    /// the smoke `dispatch-job.sh` requester doesn't publish a real
    /// job record; the console (which does) populates it. When
    /// missing, the provider can still serve the request but cannot
    /// publish a receipt — the receipt format requires both halves.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub job_cid: Option<String>,
    pub requester_did: String,
    pub requester_pub_key: String, // base64 X25519
    pub model: String,
    pub max_tokens_out: u32,
    pub ciphertext: Vec<u8>, // sealed plaintext prompt, base64-decoded by the wire layer
    /// How to interpret the opened `ciphertext` bytes. Absent (legacy) or
    /// `"text"`: a raw flattened prompt string. `"messages-v1"`: the
    /// canonical multimodal envelope (text + inline images). Additive —
    /// old consoles omit it and the provider treats that as `text`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub input_format: Option<String>,
    pub session_id: String,
    /// Confidential tier: the requester's fresh nonce. When present, the
    /// provider mints a per-request ephemeral key and returns a `SessionKey`
    /// signed over this nonce before serving. Additive — absent = best-effort.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub nonce: Option<String>,
    /// Confidential tier: CID of the attestation the requester verified, which
    /// the provider's `SessionKey` signature binds to (so a key signed for a
    /// different attestation can't be replayed). Additive.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub attestation_cid: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionKey {
    pub session_id: String,
    /// base64 X25519 public key the requester seals the prompt to.
    pub ephemeral_pub_key: String,
    /// Echo of the requester's fresh nonce.
    pub nonce: String,
    /// CID of the attestation this session is bound to.
    pub attestation_cid: String,
    /// Secure Enclave P-256 signature (DER) over the canonical bytes of
    /// `{attestationCid, ephemeralPubKey, nonce}` — the exact form the SDK's
    /// `sessionKeyMessage` reconstructs and verifies against
    /// `attestation.publicKey`.
    pub signature: Vec<u8>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodeAttestationResponse {
    /// The plaintext nonce recovered by decrypting the APNs-delivered
    /// challenge with the provider's X25519 key `K`. Echoing it proves the
    /// responder holds `K` (and, because only the genuine binary could receive
    /// the AMFI-gated push, that the responder *is* that binary).
    pub nonce: String,
    /// Secure Enclave P-256 signature (DER) over the canonical bytes of
    /// `{ nonce }` — the same canonicalisation the advisor's verifier
    /// reconstructs. Binds the recovered nonce to the attested signing key so
    /// the push receipt and the attestation chain are the same machine.
    pub signature: Vec<u8>,
}

/// Which logical channel a streamed chunk's plaintext belongs to. Lets a
/// requester separate a thinking model's reasoning from the answer. Absent
/// on the wire (old providers) deserializes to [`Content`](ChunkChannel::Content),
/// so the field is purely additive.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ChunkChannel {
    #[default]
    Content,
    Reasoning,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InferenceChunk {
    pub session_id: String,
    pub seq: u32,
    #[serde(default, skip_serializing_if = "is_content_channel")]
    pub channel: ChunkChannel,
    pub ciphertext: Vec<u8>,
}

/// Skip serializing the default `content` channel so existing wire bytes for
/// answer chunks are unchanged; only `reasoning` chunks carry the field.
fn is_content_channel(c: &ChunkChannel) -> bool {
    matches!(c, ChunkChannel::Content)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InferenceKeepalive {
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InferenceComplete {
    pub session_id: String,
    pub tokens_in: u32,
    pub tokens_out: u32,
    pub receipt_uri: String, // at:// where the receipt was published
    /// Inclusion pointer for the published receipt: the signed repo
    /// commit `rev` (and its CID) the receipt landed in. Lets the
    /// requester locate the receipt in the provider's signed MST without
    /// re-walking the repo. Empty/`None` when no receipt was published.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub receipt_commit_rev: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub receipt_commit_cid: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttestationChallenge {
    pub nonce: String,
    /// Advisor-supplied wall-clock at the moment the challenge was
    /// sent. The provider signs this verbatim into its response so the
    /// advisor can prove the response was produced after time T, not
    /// replayed from an earlier exchange.
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttestationResponse {
    pub nonce: String,
    /// Echo of the coordinator-supplied timestamp from the challenge.
    /// Signed alongside `nonce` and `sip_enabled`.
    pub timestamp: chrono::DateTime<chrono::Utc>,
    pub sip_enabled: bool,
    /// Optional: hypervisor presence as detected by the provider.
    /// `None` means the provider declined to claim either way (e.g.
    /// running on a platform where the check is unreliable).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hypervisor_present: Option<bool>,
    pub signature: Vec<u8>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn inference_chunk_channel_defaults_to_content_when_absent() {
        // An old provider sends no `channel` field; it must deserialize as the
        // answer channel so existing peers keep working.
        let json = r#"{"session_id":"s","seq":0,"ciphertext":[1,2,3]}"#;
        let chunk: InferenceChunk = serde_json::from_str(json).unwrap();
        assert_eq!(chunk.channel, ChunkChannel::Content);
    }

    #[test]
    fn content_channel_is_omitted_but_reasoning_is_serialized() {
        let content = InferenceChunk {
            session_id: "s".into(),
            seq: 0,
            channel: ChunkChannel::Content,
            ciphertext: vec![1],
        };
        let s = serde_json::to_string(&content).unwrap();
        assert!(
            !s.contains("channel"),
            "content channel should be omitted: {s}"
        );

        let reasoning = InferenceChunk {
            channel: ChunkChannel::Reasoning,
            ..content
        };
        let s = serde_json::to_string(&reasoning).unwrap();
        assert!(s.contains("\"channel\":\"reasoning\""), "got: {s}");
        // And it round-trips back.
        let back: InferenceChunk = serde_json::from_str(&s).unwrap();
        assert_eq!(back.channel, ChunkChannel::Reasoning);
    }
}

File: infra/advisor/src/protocol.ts
// Wire types for the advisor ↔ provider WebSocket. Mirrors
// provider/src/protocol.rs. Field naming is the Rust struct field
// name verbatim (snake_case for fields like `sip_enabled`); the
// `type` discriminator is snake_case too because the Rust enum
// has `#[serde(tag = "type", rename_all = "snake_case")]`.
//
// Bytes (Vec<u8> on the Rust side) are emitted by serde_json as a
// JSON array of integers. We accept both that shape and a base64
// string so a future wire change to base64 doesn't break us.

/** Diagnostic the provider sends when it couldn't bring one or more
 *  configured inference engines online after its startup recovery loop
 *  gave up. The failed models are already absent from
 *  {@link Register.supported_models}, so this does not affect matchmaking
 *  — it lets the advisor note (log + expose) that a connected machine is
 *  degraded (only serving `stub`). Mirror of the provider's
 *  `dev.cocore.compute.provider#engineFault`. */
export interface EngineFault {
  code: string;
  message: string;
  models?: string[];
  at: string;
}

export interface Register {
  provider_did: string;
  /** Stable per-machine identifier — the agent's `dev.cocore.compute.provider`
   *  record rkey. Distinguishes two machines that register under the same DID
   *  so the advisor holds both instead of evicting one. Optional / additive:
   *  pre-`machine_id` agents omit it and the advisor falls back to the
   *  per-machine `attestation_pub_key` (see registry.machineIdOf). */
  machine_id?: string;
  machine_label: string;
  chip: string;
  ram_gb: number;
  supported_models: string[];
  encryption_pub_key: string;
  attestation_pub_key: string;
  attestation_uri: string;
  /** Optional / additive — absent on a healthy serve. */
  engine_fault?: EngineFault;
  /** Measured cdHash echoed from the signed attestation — the advisor checks
   *  it against its known-good set when computing confidential eligibility.
   *  Additive. */
  cd_hash?: string;
  /** Provider's self-asserted tier (`attested-confidential` | `best-effort`),
   *  echoed from the signed attestation. Advisory — the advisor recomputes.
   *  Additive. */
  tier?: string;
  /** Coarse, opt-in ISO 3166-1 alpha-2 country echoed from the provider
   *  record's `region`, so the advisor's /providers list can route by
   *  country without a PDS read. Advisory self-claim; absent when the owner
   *  hasn't opted into location sharing. Additive. */
  region?: string;
  /** APNs device token (hex) for the measured agent process, when it could
   *  register for remote notifications (confidential build + GUI session). The
   *  advisor sends the code-identity challenge here. Omitted on headless
   *  installs, which therefore stay best-effort. Additive. */
  apns_device_token?: string;
  /** True only when at least one model passed the provider's startup
   *  forced-tool canary. `tool_call_models` carries the per-model verified
   *  subset for new clients. Additive — old advisors ignore it. */
  supports_tool_calls?: boolean;
  /** Model ids whose engines passed the provider's forced-tool startup canary.
   *  When absent, clients fall back to legacy `supports_tool_calls`; when
   *  present, tool-capability gating should require the requested model to be
   *  listed. Additive. */
  tool_call_models?: string[];
  /** This agent binary's version (e.g. `0.9.32`), echoed live so the advisor
   *  can route version-gated jobs (a request needing a feature only present
   *  from some release is steered to machines at/above a minimum version).
   *  Additive — pre-version agents omit it, and a version floor treats a
   *  machine that omits it as below the floor (fail-closed). */
  binary_version?: string;
  /** True when the agent's signing key is Secure-Enclave-resident (ADR-0005
   *  confidential-tier evidence, echoed from `attestation.secureEnclaveAvailable`).
   *  Additive: old software-key agents omit it; the SE gate (when enforced)
   *  treats omitted/false as not-SE → best-effort. */
  secure_enclave_available?: boolean;
  /** The agent's encryption-key scheme: `"p256-ecies-se"` (Secure-Enclave key)
   *  or absent/`"x25519"` (software). Tells the advisor which codec to seal the
   *  APNs code-challenge nonce with, so an old X25519 agent keeps working.
   *  Additive. */
  enc_scheme?: string;
  /** atproto service-auth JWT that binds this Register to `provider_did`.
   *  The provider mints it via `com.atproto.server.getServiceAuth` with
   *  `aud = COCORE_ADVISOR_DID` and `lxm = "dev.cocore.compute.register"`; its
   *  PDS signs it with the DID's repo signing key. The advisor verifies it and
   *  requires the authenticated DID to equal `provider_did` before accepting
   *  the registration — without it a client could register as any provider and
   *  swap in its own attestation key. Additive: optional on the wire so the
   *  fleet can upgrade before the advisor flips `COCORE_ADVISOR_REQUIRE_AUTH`
   *  on (see connection.ts). */
  auth_jwt?: string;
}

/** Content-free crash signature the provider folds into its heartbeat
 *  after its inference worker panicked / aborted and was restarted. It
 *  carries no payload — just enough for the advisor to spot a machine
 *  that's crash-looping (high `count` in a short window) and stop feeding
 *  it jobs. Mirror of the provider's `Heartbeat.crash` (serde, snake_case
 *  field names). All fields but `count` are omitted when absent; a machine
 *  that never crashed omits the whole object. Additive — old agents never
 *  send it. */
export interface CrashSignature {
  count: number;
  last_at?: string;
  location?: string;
  signature?: string;
  version?: string;
}

interface Heartbeat {
  load: number;
  queue_depth: number;
  at: string;
  /** The owner's start/stop switch as the agent sees it on its PDS record.
   *  `false` → the owner stopped this machine; the advisor routes it no
   *  jobs. `true` / absent → serving. Additive. */
  active?: boolean;
  /** Latest crash signature, present only once the provider has crashed at
   *  least once this session. Absent on a machine that's never crashed.
   *  Additive — old agents omit it. */
  crash?: CrashSignature;
}

export interface AttestationChallenge {
  nonce: string;
  timestamp: string;
}

export interface AttestationResponse {
  nonce: string;
  timestamp: string;
  sip_enabled: boolean;
  hypervisor_present?: boolean;
  signature: number[] | string;
}

/** Provider → advisor: response to an APNs code-identity challenge. The
 *  challenge arrives out-of-band over APNs (a nonce sealed to the provider's
 *  X25519 key `K`); only the genuine, AMFI-gated binary can receive and open
 *  it. The provider echoes the recovered `nonce` and a Secure-Enclave P-256
 *  signature (DER) over the canonical `{ nonce }`. Mirror of the Rust
 *  `CodeAttestationResponse`. Additive. */
export interface CodeAttestationResponse {
  nonce: string;
  signature: number[] | string;
}

export interface InferenceRequest {
  job_uri: string;
  /** Optional CID half of the job strong-ref. The provider needs
   *  both halves to publish a receipt. The smoke
   *  `dispatch-job.sh` requester doesn't have a real job record;
   *  the console (which does) populates this. */
  job_cid?: string;
  requester_did: string;
  requester_pub_key: string;
  model: string;
  max_tokens_out: number;
  ciphertext: number[] | string;
  /** How the provider interprets the opened ciphertext: absent/"text"
   *  (raw prompt) or "messages-v1" (multimodal envelope). Forwarded from
   *  the `/jobs` body; the advisor never reads the plaintext. */
  input_format?: string;
  session_id: string;
  /** Optional JSON Schema constraining the model's output. When present,
   *  the provider passes it to the inference engine as response_format
   *  guided decoding. Forwarded from the `/jobs` body; the advisor never
   *  reads the plaintext. */
  output_schema?: { name: string; strict?: boolean; schema: Record<string, unknown> };
  /** Optional tool definitions the model may call. Forwarded from the
   *  `/jobs` body; the advisor never inspects the plaintext. */
  tools?: unknown;
  /** Optional tool-choice directive (e.g. "auto", "none", "required", or
   *  a provider-specific object). Forwarded from the `/jobs` body; the
   *  advisor never inspects it. */
  tool_choice?: unknown;
  /** ADR-0004: the brokerage's session-bound countersignature for this
   *  dispatch, signed at routing time over {authority, job, requester, machine,
   *  attestation, nonce}. The provider copies it onto the receipt as
   *  `brokerageCountersignature` (lexicon camelCase) so a confidential requester
   *  can verify the job was routed by a trusted authority to the attested
   *  machine. Present only when the advisor has a brokerage authority key
   *  configured AND the job carries the fields needed to bind it. Additive. */
  brokerage_countersignature?: {
    authority: string;
    machine_id: string;
    nonce: string;
    sig: string;
  };
}

interface InferenceChunk {
  session_id: string;
  seq: number;
  /** Which channel this chunk's plaintext belongs to. Absent (older
   *  providers) means the answer. The advisor relays it opaquely. */
  channel?: "content" | "reasoning" | "tool_call";
  ciphertext: number[] | string;
}

interface InferenceComplete {
  session_id: string;
  tokens_in: number;
  tokens_out: number;
  receipt_uri: string;
}

/** Provider → advisor: "still generating" signal sent during a long job
 *  when no user-visible token has gone out for a while (slow prefill or a
 *  slow decode patch). Resets the session idle timer without being a token
 *  — keeps a slow-but-alive job from being killed as silent. Additive; an
 *  advisor that doesn't know this frame simply ignores it. */
interface InferenceKeepalive {
  session_id: string;
}

/** Advisor → provider liveness probe. The provider answers `pong` with
 *  the same nonce IMMEDIATELY from its serve loop (not behind an
 *  in-flight inference), so a pong proves the agent's request loop is
 *  actually pumping — a stronger signal than a WS-level pong, which the
 *  read half auto-sends even when the serve loop is wedged. Used by
 *  `/jobs` to preflight the chosen provider before committing a job. */
interface Ping {
  nonce: string;
}

/** Provider → advisor reply to {@link Ping}, echoing the nonce. */
interface Pong {
  nonce: string;
}

/** Advisor → provider standing change. `bad` means the advisor stopped
 *  routing jobs to this machine (it failed a preflight ping or went
 *  silent mid-job); the agent surfaces it to the operator (red tray
 *  ping). `ok` clears it. Additive — old agents ignore it. */
interface HealthNotice {
  standing: "ok" | "bad";
  reason?: string;
}

/** Advisor → provider nudge: "your owner-set control state changed —
 *  re-read it." Carries no authority; the agent re-reads the authoritative
 *  `active` from its own PDS and reports back in a heartbeat. Lets a
 *  console start/stop take effect in ~a second instead of at the next
 *  poll. Additive. */
interface ControlChanged {
  reason?: string;
}

/** Advisor → provider: "you've been flagged unhealthy — try to self-right
 *  NOW." The agent runs its in-process recovery (engine health-check +
 *  bounded restart, re-attest) immediately rather than waiting for its next
 *  scheduled health tick, then reports the outcome in {@link RecoverResult}.
 *  Sent automatically the moment the advisor marks a machine unhealthy, and
 *  on demand when the owner clicks "Try to recover" in the console. Carries
 *  no authority. Additive — old agents ignore it (they still get the
 *  `health_notice` that accompanies it). */
interface RecoverRequest {
  reason?: string;
}

/** Provider → advisor: outcome of a self-right attempt. `recovered: true`
 *  means the machine brought its engine(s) back and is ready for jobs again
 *  (the advisor clears its bad standing); `false` means it couldn't, and
 *  `detail` carries the human-readable reason the console + tray surface as
 *  remediation. Additive. */
interface RecoverResult {
  recovered: boolean;
  detail?: string;
}

export type AdvisorMessage =
  | ({ type: "register" } & Register)
  | ({ type: "heartbeat" } & Heartbeat)
  | ({ type: "attestation_challenge" } & AttestationChallenge)
  | ({ type: "attestation_response" } & AttestationResponse)
  | ({ type: "code_attestation_response" } & CodeAttestationResponse)
  | ({ type: "inference_request" } & InferenceRequest)
  | ({ type: "inference_chunk" } & InferenceChunk)
  | ({ type: "inference_keepalive" } & InferenceKeepalive)
  | ({ type: "inference_complete" } & InferenceComplete)
  | ({ type: "ping" } & Ping)
  | ({ type: "pong" } & Pong)
  | ({ type: "health_notice" } & HealthNotice)
  | ({ type: "control_changed" } & ControlChanged)
  | ({ type: "recover_request" } & RecoverRequest)
  | ({ type: "recover_result" } & RecoverResult);

// --- Inbound frame validation --------------------------------------
// `JSON.parse` proves a frame is JSON; it proves nothing about its shape.
// The handlers dereference per-type fields directly (e.g.
// `msg.supported_models.length`), so a frame like `{"type":"register"}` — or
// an `attestation_response` missing `signature` — would throw a TypeError deep
// in a handler, surface as an unhandled rejection, and kill the process. We
// validate the shape of the fields each handler touches BEFORE dispatch and
// close the socket cleanly on a mismatch instead. Lightweight and hand-written
// (no schema dep): it only checks the fields the advisor actually reads, and
// stays additive (unknown fields and unknown `type`s are tolerated — an
// advisor that doesn't know a frame simply ignores it).

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}
const isStr = (v: unknown): v is string => typeof v === "string";
/** Bytes on the wire: a base64 string OR a JSON array of byte values. */
const isBytes = (v: unknown): boolean =>
  typeof v === "string" || (Array.isArray(v) && v.every((n) => typeof n === "number"));

/** Result of validating a raw inbound frame: either the typed message or a
 *  reason string for the `close(1008,"bad-frame")` log. */
export type FrameCheck = { ok: true; msg: AdvisorMessage } | { ok: false; reason: string };

/** Validate a JSON-parsed inbound frame against the fields the advisor's
 *  handlers dereference. Only provider→advisor frames are checked; advisor→
 *  provider frames (which the advisor never receives) fall through as accepted
 *  so an unexpected echo is ignored rather than closing the socket. */
export function validateFrame(raw: unknown): FrameCheck {
  if (!isRecord(raw)) return { ok: false, reason: "frame is not a JSON object" };
  const type = raw["type"];
  if (!isStr(type)) return { ok: false, reason: "missing string `type`" };
  switch (type) {
    case "register": {
      if (!isStr(raw["provider_did"])) return { ok: false, reason: "register: provider_did" };
      if (!Array.isArray(raw["supported_models"])) {
        return { ok: false, reason: "register: supported_models" };
      }
      if (!isStr(raw["encryption_pub_key"]))
        return { ok: false, reason: "register: encryption_pub_key" };
      if (!isStr(raw["attestation_pub_key"]))
        return { ok: false, reason: "register: attestation_pub_key" };
      break;
    }
    case "attestation_response": {
      if (!isStr(raw["nonce"])) return { ok: false, reason: "attestation_response: nonce" };
      // `signature` is fed to bytesToBase64 during verify; require it present
      // and byte-shaped so verify can't throw on undefined.
      if (!isBytes(raw["signature"]))
        return { ok: false, reason: "attestation_response: signature" };
      break;
    }
    case "code_attestation_response": {
      if (!isStr(raw["nonce"])) return { ok: false, reason: "code_attestation_response: nonce" };
      if (!isBytes(raw["signature"]))
        return { ok: false, reason: "code_attestation_response: signature" };
      break;
    }
    case "inference_chunk": {
      if (!isStr(raw["session_id"])) return { ok: false, reason: "inference_chunk: session_id" };
      if (!isBytes(raw["ciphertext"])) return { ok: false, reason: "inference_chunk: ciphertext" };
      break;
    }
    case "inference_keepalive": {
      if (!isStr(raw["session_id"]))
        return { ok: false, reason: "inference_keepalive: session_id" };
      break;
    }
    case "inference_complete": {
      if (!isStr(raw["session_id"])) return { ok: false, reason: "inference_complete: session_id" };
      break;
    }
    case "pong": {
      if (!isStr(raw["nonce"])) return { ok: false, reason: "pong: nonce" };
      break;
    }
    // heartbeat / recover_result carry only optional fields the handlers
    // guard individually; the `type` check above is enough. Advisor→provider
    // frame `type`s (and any unknown `type`) fall through as accepted and are
    // ignored by onMessage's switch.
    default:
      break;
  }
  return { ok: true, msg: raw as unknown as AdvisorMessage };
}

export function bytesToBase64(b: number[] | string): string {
  if (typeof b === "string") return b;
  // Defend against a malformed frame handing us a non-array (e.g. an
  // `attestation_response` with `signature` missing → `undefined`): fail with a
  // clear, catchable error instead of throwing deep inside the for-of loop with
  // an opaque message. The message-dispatch wrapper turns this into a clean
  // `close(1008,"bad-frame")` rather than a process-killing rejection.
  if (!Array.isArray(b)) {
    throw new TypeError("bytesToBase64: expected number[] or base64 string");
  }
  let bin = "";
  for (const byte of b) bin += String.fromCharCode(byte);
  return btoa(bin);
}
